### PR TITLE
Add support for 1.21.1

### PR DESCRIPTION
**Pull request by Gilly7CE**
This PR introduces a support change for Minecraft (MC) 1.21.1. This directly bumps main from 1.20.4 to 1.21.1 so it includes fixes to 1.20.5 & 1.20.6 breaking changes. A PR to update the 1.20.x branch to 1.20.6 will be done separately.

Full list of changes:
- Update Gradle to 8.10 and Loom to 1.7-Snapshot
- Update MC support to 1.21.1
- Fix breaking change to `BlockBreakingSpeedRatioCalculator` following enchantment code changes in MC
- Replace `EndFrameBlock_EyeToggleMixin` with `EnderEyeItem_EyeToggleMixin` following removal of deprecated methods in `Block` code
- Fix mixin definitions following MC code changes
- Replace mixins for flower pots with single mixin and introduce functional approach to dealing with which rule to implement
- Replace `PlayerEntity_PortalSpectactorDelayMixin` with `NetherPortalBlock_PortalBlockMixin` following portal interaction changes
- Refactor spectator portal rule code to not use networking sync between server and client but two mixins for server and client which executes the same logic, like in vanilla. Networking changes in 1.20.5 caused a breaking change here and it would be too complicated to keep around for a single use case. The new implementation is likely more future proof.

### CODE REVIEW

**Comment by Gilly7CE** [comment]
Keeping in draft until pipeline issue is fixed and I've tested all rules affected by breaking changes

**Comment by Gilly7CE** [comment]
> Keeping in draft until pipeline issue is fixed and I've tested all rules affected by breaking changes

Fixed one issue to do with instant mining where the game crashed

Two rules now broken:

- [ ] `replaceFlowerInPot`
- [ ] `dropAllXpOnDeath`

**Comment by Gilly7CE** [comment]
> > Keeping in draft until pipeline issue is fixed and I've tested all rules affected by breaking changes
> 
> Fixed one issue to do with instant mining where the game crashed
> 
> Two rules to fix:
> 
> * [x]  `replaceFlowerInPot`
> * [x]  `alwaysPickFlowersFromPot`
> * [x]  `dropAllXpOnDeath`

These rules have now been fixed. Introduced a single mixin for the flower pot rules to manage the order instead of using mixin shifts. This will be more stable.

File: src/main/java/carpetaddonsnotfound/mixins/LivingEntity_DropAllXpOnPlayerDeathMixin.java
Comment (connorslade) [inline src/main/java/carpetaddonsnotfound/mixins/LivingEntity_DropAllXpOnPlayerDeathMixin.java:1]: From testing this doesn't seem to work any more.

**Review by connorslade — CHANGES_REQUESTED** [review]
Looks like just DropAllXpOnPlayerDeath needs to be fixed.

File: src/main/java/carpetaddonsnotfound/mixins/LivingEntity_DropAllXpOnPlayerDeathMixin.java
Comment (Gilly7CE) [reply src/main/java/carpetaddonsnotfound/mixins/LivingEntity_DropAllXpOnPlayerDeathMixin.java:1]: Looks like the formulas are still valid. Not sure what I did that had this rule no longer working following the mixin method inject definition update...restored the formulas in the latest update.

**Review by connorslade — APPROVED** [review]
👍

**Pull request merged**